Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/model/GameFileCache.java

```diff
@@ -28,70 +23,89 @@ public GameFileCache(String path)
   @Override
   public native void finalize();
 
-  public static void addGameFolder(String path, Context context)
+  public static void addGameFolder(String path)
   {
-    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
-    Set<String> folderPaths = preferences.getStringSet(GAME_FOLDER_PATHS_PREFERENCE, EMPTY_SET);
+    File dolphinFile = SettingsFile.getSettingsFile(Settings.FILE_DOLPHIN);
+    IniFile dolphinIni = new IniFile(dolphinFile);
+    LinkedHashSet<String> pathSet = getPathSet(false);
+    int totalISOPaths =
+            dolphinIni.getInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS, 0);
 
-    if (folderPaths == null)
+    if (!pathSet.contains(path))
     {
-      return;
+      dolphinIni.setInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS,
+              totalISOPaths + 1);
+      dolphinIni.setString(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATH_BASE +
+              totalISOPaths, path);
+      dolphinIni.save(dolphinFile);
+      NativeLibrary.ReloadConfig();
     }
-
-    Set<String> newFolderPaths = new HashSet<>(folderPaths);
-    newFolderPaths.add(path);
-    SharedPreferences.Editor editor = preferences.edit();
-    editor.putStringSet(GAME_FOLDER_PATHS_PREFERENCE, newFolderPaths);
-    editor.apply();
   }
 
-  private void removeNonExistentGameFolders(Context context)
+  private static LinkedHashSet<String> getPathSet(boolean removeNonExistentFolders)
   {
-    SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
-    Set<String> folderPaths = preferences.getStringSet(GAME_FOLDER_PATHS_PREFERENCE, EMPTY_SET);
+    File dolphinFile = SettingsFile.getSettingsFile(Settings.FILE_DOLPHIN);
+    IniFile dolphinIni = new IniFile(dolphinFile);
+    LinkedHashSet<String> pathSet = new LinkedHashSet<>();
+    int totalISOPaths =
+            dolphinIni.getInt(Settings.SECTION_INI_GENERAL, SettingsFile.KEY_ISO_PATHS, 0);
 
-    if (folderPaths == null)
+    if (totalISOPaths < 1)
     {
-      return;
+      return pathSet;
     }
 
-    Set<String> newFolderPaths = new HashSet<>();
-    for (String folderPath : folderPaths)
+    for (int i = 0; i < totalISOPaths; i++)
     {
-      File folder = new File(folderPath);
+      String path = dolphinIni.getString(Settings.SECTION_INI_GENERAL,
+              SettingsFile.KEY_ISO_PATH_BASE + i, "");
+
+      File folder = new File(path);
       if (folder.exists())
```

Comment:
What do you think about changing this to `if (!removeNonExistentFolders || folder.exists())`? It seems like `addGameFolder` is expecting the `getPathSet(false)` call to return everything that's in the INI file.

Reply:
If folders that don't exist are added to the set when `addGameFolder` is called, won't that be unnecessary items for `!pathSet.contains(path)` to search through?

Reply:
It depends on what we consider unnecessary. The only case where the distinction should matter is when we try to add a folder that the system considers not to exist. If that is a case we care about, then folders that don't exist need to be in `pathSet` so that we don't get duplicates. It would probably be pretty hard to trigger this case in practice with the ways `addGameFolder` currently are called, though...

Reply:
`FilePicker` (in `MODE_DIR`) only allows folders to be passed to `addGameFolder` so I think we shouldn't doubt the existence of the to-be-added folder. Unless the system that will replace `FilePicker` for shared storage doesn't check a folder's existence before returning it, I still don't see a reason to not minimize the size of the set.

Reply:
Yeah, something really unlikely would have to happen, like the folder getting deleted right after the file picker is done running. I was thinking more about if we decide to call `addGameFolder` in new ways in the future. Whoever makes such a change in some time from now probably won't think of fixing this problem, since finding out about it requires reading the implementation of `addGameFolder` fairly closely.

Reply:
Even if that was the case, `scanLibrary` is called after `addGameFolder` (in addition to MainActivity's `onResume`). Non-existent files in `ISOPath#` won't survive very long. Correct me if I'm making a bad design decision here but this just seems like a scenario that's adequately handled elsewhere.

Reply:
Fair enough then.